### PR TITLE
Enhance forRange to accept negative increments (#5520)

### LINKED ISSUE #5520: Enhance `forRange` to accept negative increments
`forRange` allows iteration based on a defined range of numbers and an increment, but currently only works if the increment is positive. Current examples:
```
forRange(0,10,1,i,i) -> [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ]
forRange(10,0,-1,i,i) -> [  ]
```

This is different to the `range()` function where a negative increment is accepted:
```
range(0,10,1) -> [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ]
range(10,0,-1) -> [ 10, 9, 8, 7, 6, 5, 4, 3, 2, 1 ]
```

### Proposed solution
forRange could be enhanced to work with a negative increment
It might also be a good idea to reject invalid arguments - for example currently:
```
forRange(10,0,1,i,i)
```
As there is no way of incrementing by 1 to get from 10 to zero, this should be rejected as invalid parameters, whereas instead it currently returns an empty array

### Alternatives considered
It is possible to use `range` combined with `forEach`:
```
forEach(range(10,0,-1),v,v) -> [ 10, 9, 8, 7, 6, 5, 4, 3, 2, 1 ]
```
This isn't hugely more complicated but it seems that `range` and `forRange` should be consistent with each other

### PR DESCRIPTION
**Pull request by Huishin-pie**
Fixes #5520

Changes proposed in this pull request:
- forRange can work with a negative increment.
- Fixes the bug to support both integer and floating point values.
- To reject invalid arguments and return an error message instead of an empty array.

### CODE REVIEW

File: main/src/com/google/refine/grel/controls/ForRange.java
Comment (tfmorris) [inline main/src/com/google/refine/grel/controls/ForRange.java:86]: We want to make sure to check for `step == 0` here since that will cause an infinite loop, but I don't think we should be more permissive about parameters which will cause zero iterations e.g. from 0 to 10 by -1 because those values might be calculated and we don't want the callers to have to do special handling when zero iterations is an adequate solution.

File: main/src/com/google/refine/grel/controls/ForRange.java
Comment (tfmorris) [inline main/src/com/google/refine/grel/controls/ForRange.java:99]: Great! Thanks for fixing this bug.

File: main/src/com/google/refine/grel/controls/ForRange.java
Comment (tfmorris) [inline main/src/com/google/refine/grel/controls/ForRange.java:93]: I think the way that this was originally organized without the intermediate `rangeSize` calculation is more direct and avoids any additional loss of precision.

File: main/src/com/google/refine/grel/controls/ForRange.java
Comment (tfmorris) [inline main/src/com/google/refine/grel/controls/ForRange.java:106]: See comments above for the integer branch.

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (tfmorris) [inline main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:69]: The last one should be an error, but all the others should succeed with zero iterations.

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (tfmorris) [inline main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:55]: Please add a test or two for things like `from == to`, `step > from - to`, and other edge cases. The code is pretty simple, so but it's a good habit to get into to test the corners.

**Review by tfmorris — CHANGES_REQUESTED** [review]
Great start! I made a couple of comments for minor things to improve.
@wetneb may have additional comments.

**Comment by Huishin-pie** [comment]
@tfmorris Thank you for the review. I will improve the code as soon as possible. :)

**Review by wetneb — APPROVED** [review]
Thanks for working on this! Nothing to add on my side beyond @tfmorris' review

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (Huishin-pie) [inline main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:83]: There was a failure in this test due to floating-point precision. Do we need to resolve this issue?

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (Huishin-pie) [inline main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:82]: There was a failure in this test due to floating-point precision. Do we need to resolve this issue?

**Comment by Huishin-pie** [comment]
Please review the changes made. Thank you!

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (tfmorris) [reply main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:82]: Sorry for missing this question! Yes, all tests need to pass before we can merge. The traditional way to do floating point comparisons is to make sure two values are within a small value of each other ie `abs(f1 - f2) < epsilon` rather than `f1 == f2` but since these are string comparisons, a different solution will be needed.

File: main/src/com/google/refine/grel/controls/ForRange.java
Comment (tfmorris) [inline main/src/com/google/refine/grel/controls/ForRange.java:90]: This test is unnecessary. It will be handled naturally by the loop logic (ie if step > (from-to) there will be zero iterations.

File: main/src/com/google/refine/grel/controls/ForRange.java
Comment (tfmorris) [inline main/src/com/google/refine/grel/controls/ForRange.java:91]: `step` is invariant here, so I would either convert this an `if` statement
```
if (step > 0) {
  while (from < to) {
  }
}
else {
  while (from < to) {
  }
}
```
or, to reduce the redundancy, fold both loops together and make the conditional slightly more complex
```
while ((step > 0 && from < to) || to > from) {
```
(we don't need to test `step < 0` because we know it's not zero and not greater than zero)

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (tfmorris) [reply main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:82]: Also, I hadn't noticed that the test workflows required approval to run. Now that they've been approved once, I'm hoping they'll run automatically on future commits.

**Review by tfmorris — CHANGES_REQUESTED** [review]

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (Huishin-pie) [reply main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:82]: Thank you for your response. This is my first time writing a test for floating-point numbers, so I'm not very familiar with it. Could you please give me some more advice? I've researched and only found that the solution as you said is to check if the expected and actual values are equal within a certain delta tolerance.

Comment (tfmorris) [reply main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:82]: Since this is operating in the string domain, one option would be to use the precision specifier on `toString()` so change the test to something like 
```
forRange(0,0.6,0.1,v,v.toString('%.3f').join(',')
```
and then adjust to expected values to something like `0.000`. This will give you enough precision to catch real errors, but not so much precision that you're bothered by differences too small to care about.

BTW, I'll be offline for the next 2-3 of days, but will check back in when I return (and @wetneb ) and the other developers can help in the mean time.

File: main/src/com/google/refine/grel/controls/ForRange.java
Comment (Huishin-pie) [reply main/src/com/google/refine/grel/controls/ForRange.java]: I misunderstood the meaning of zero iterations. I have adjusted the code.

File: main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java
Comment (Huishin-pie) [reply main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:82]: I got it! Thanks for your help.

Comment (tfmorris) [reply main/tests/server/src/com/google/refine/grel/controls/ForRangeTests.java:82]: Perfect!

**Review by tfmorris — APPROVED** [review]
Looks great. Thanks for sticking with it!

**Pull request merged**
